# Alocação e Otimização em Portfólio de Ativos

Importando bibliotecas Python:

In [1]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt 
import plotly.express as px
import yfinance as yf 

In [2]:
acao = ['PETR4.SA', 'VALE3.SA', 'BBDC4.SA', 'RENT3.SA', 'BRFS3.SA','BOVA11.SA']

In [3]:
dataset = pd.DataFrame()
for acoes in acao:
    dataset[acoes] = yf.download(acoes, start='2015-01-01', end='2020-11-04')['Close']

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed


In [4]:
dataset

PETR4.SA   VALE3.SA   BBDC4.SA   RENT3.SA   BRFS3.SA  BOVA11.SA
Date                                                                        
2015-01-02   9.360000  21.280001  13.445455  10.568521  59.189453  47.259998
2015-01-05   8.560000  20.959999  13.465055  10.129169  58.066204  46.320000
2015-01-06   8.280000  21.799999  13.908010  10.333799  58.589752  46.580002
2015-01-07   8.670000  22.600000  14.460724  10.442132  61.207493  48.150002
2015-01-08   9.230000  22.840000  14.535203  10.264586  60.112801  48.509998
...               ...        ...        ...        ...        ...        ...
2020-10-27  19.879999  62.529999  18.396694  64.267342  16.591703  95.680000
2020-10-28  18.670000  60.259998  17.355371  62.256813  16.182384  91.559998
2020-10-29  19.290001  62.020000  17.181818  62.266766  16.106232  92.800003
2020-10-30  18.940001  60.549999  16.661158  60.395576  15.972965  90.660004
2020-11-03  19.650000  63.340000  17.231403  63.471096  16.258537  92.260002

[1453 rows x 6 columns]

Verificando e eliminando valores nulos:

In [5]:
dataset.isnull().sum()

PETR4.SA      0
VALE3.SA      0
BBDC4.SA      0
RENT3.SA      0
BRFS3.SA      0
BOVA11.SA    27
dtype: int64

In [6]:
dataset.dropna(inplace=True)
dataset.isnull().sum()

PETR4.SA     0
VALE3.SA     0
BBDC4.SA     0
RENT3.SA     0
BRFS3.SA     0
BOVA11.SA    0
dtype: int64

Renomeando as colunas:

In [7]:
dataset = dataset.rename(columns={'PETR4.SA': 'PETROBRAS', 'VALE3.SA': 'VALE', 'BBDC4.SA': 'BRADESCO', 'RENT3.SA': 'RENTACAR', 'BRFS3.SA': 'BRFOOD', 'BOVA11.SA': 'BOVA'})

Criando e lendo arquivo .csv:

In [8]:
dataset.to_csv('acoes_novas.csv')


In [9]:
dataset = pd.read_csv('acoes_novas.csv')

In [10]:
dataset

Date  PETROBRAS       VALE   BRADESCO   RENTACAR     BRFOOD  \
0     2015-01-02   9.360000  21.280001  13.445455  10.568521  59.189453   
1     2015-01-05   8.560000  20.959999  13.465055  10.129169  58.066204   
2     2015-01-06   8.280000  21.799999  13.908010  10.333799  58.589752   
3     2015-01-07   8.670000  22.600000  14.460724  10.442132  61.207493   
4     2015-01-08   9.230000  22.840000  14.535203  10.264586  60.112801   
...          ...        ...        ...        ...        ...        ...   
1421  2020-10-27  19.879999  62.529999  18.396694  64.267342  16.591703   
1422  2020-10-28  18.670000  60.259998  17.355371  62.256813  16.182384   
1423  2020-10-29  19.290001  62.020000  17.181818  62.266766  16.106232   
1424  2020-10-30  18.940001  60.549999  16.661158  60.395576  15.972965   
1425  2020-11-03  19.650000  63.340000  17.231403  63.471096  16.258537   

           BOVA  
0     47.259998  
1     46.320000  
2     46.580002  
3     48.150002  
4     48.509998  
...         ...  
1421  95.680000  
1422  91.559998  
1423  92.800003  
1424  90.660004  
1425  92.260002  

[1426 rows x 7 columns]

Descrição do portfólio:

In [11]:
dataset.describe()

PETROBRAS         VALE     BRADESCO     RENTACAR       BRFOOD  \
count  1426.000000  1426.000000  1426.000000  1426.000000  1426.000000   
mean     17.317735    36.142468    18.101469    22.939061    37.808745   
std       6.985286    15.660930     5.168432    14.254232    14.977687   
min       4.200000     8.600000     8.001417     6.165973    12.384284   
25%      12.210000    19.767500    14.456653    10.586576    22.226980   
50%      15.900000    35.344999    17.138962    19.640838    36.648342   
75%      22.877500    50.434999    21.758959    32.485112    49.513337   
max      30.969999    63.450001    29.609316    64.267342    68.603790   

              BOVA  
count  1426.000000  
mean     71.662896  
std      19.970061  
min      36.450001  
25%      53.272500  
50%      70.805000  
75%      90.565001  
max     115.209999

# Análise Financeira

## Resumo Estatístico

A tabela abaixo apresenta um resumo estatístico dos dados financeiros das empresas listadas:

|           | PETROBRAS | VALE      | BRADESCO  | RENTACAR  | BRFOOD    | BOVA      |
|-----------|-----------|-----------|-----------|-----------|-----------|-----------|
| count     | 1426.000  | 1426.000  | 1426.000  | 1426.000  | 1426.000  | 1426.000  |
| mean      | 17.318    | 36.142    | 18.101    | 22.939    | 37.809    | 71.663    |
| std       | 6.985     | 15.661    | 5.168     | 14.254    | 14.978    | 19.970    |
| min       | 4.200     | 8.600     | 8.001     | 6.166     | 12.384    | 36.450    |
| 25%       | 12.210    | 19.768    | 14.457    | 10.587    | 22.227    | 53.273    |
| 50%       | 15.900    | 35.345    | 17.139    | 19.641    | 36.648    | 70.805    |
| 75%       | 22.878    | 50.435    | 21.759    | 32.485    | 49.513    | 90.565    |
| max       | 30.970    | 63.450    | 29.609    | 64.267    | 68.604    | 115.210   |

## Observações

- **Count**: O número total de observações disponíveis para cada empresa é consistente em 1426.
- **Média (Mean)**: Representa o valor médio das ações ao longo do período analisado para cada empresa. Por exemplo, a média das ações da Petrobras é de aproximadamente 17.32.
- **Desvio Padrão (Std)**: Indica o grau de dispersão dos dados em relação à média. Quanto maior o desvio padrão, maior é a variabilidade dos valores. Por exemplo, as ações da BRFOOD têm um desvio padrão de aproximadamente 14.98, indicando uma alta variabilidade em comparação com outras empresas.
- **Mínimo (Min)**: O valor mínimo observado para as ações de cada empresa durante o período.
- **Máximo (Max)**: O valor máximo observado para as ações de cada empresa durante o período.
- **Percentis (25%, 50%, 75%)**: Representam os valores que dividem os dados em quatro partes iguais. Por exemplo, o percentil de 25% (primeiro quartil) para a VALE é de aproximadamente 19.77, o que significa que 25% das observações de preço das ações da Vale são iguais ou inferiores a esse valor.

Esses dados fornecem uma visão geral das características das ações das empresas listadas no DataFrame. 

***

Com base nas estatísticas do DataFrame fornecido, podemos tirar algumas conclusões sobre o risco e a rentabilidade das ações listadas, bem como do índice BOVA:

Risco:

- As ações com menor desvio padrão (indicando menor volatilidade e, portanto, menor risco) são PETROBRAS, BRADESCO e RENTACAR, com desvios padrão de 6.99, 5.17 e 14.25, respectivamente.
- As ações com maior desvio padrão (maior volatilidade e, portanto, maior risco) são BRFOOD, VALE e BOVA, com desvios padrão de 14.98, 15.66 e 19.97, respectivamente.

Rentabilidade:

- As ações com maior média de preço ao longo do período são BRFOOD, VALE e BOVA, com médias de 37.81, 36.14 e 71.66, respectivamente.
- As ações com menor média de preço ao longo do período são PETROBRAS, BRADESCO e RENTACAR, com médias de 17.32, 18.10 e 22.94, respectivamente.

Comparação:

- Em termos de rentabilidade, as ações da BRFOOD se destacam com a maior média de preço, seguidas por VALE e BOVA. No entanto, em relação ao risco, as ações da BRFOOD também apresentam um desvio padrão mais alto.
- Por outro lado, as ações da PETROBRAS têm a menor média de preço, mas também o menor desvio padrão, sugerindo um menor risco em relação às outras ações.
- O índice BOVA, apesar de possuir uma média de preço intermediária, tem uma volatilidade comparável às das ações individuais.

Conclusão:

Considerando tanto o risco quanto a rentabilidade, as ações da PETROBRAS se destacam como uma opção com menor risco e rentabilidade comparável.
Para investidores com maior apetite ao risco, as ações da BRFOOD e VALE podem oferecer oportunidades de maior retorno, mas com uma volatilidade mais elevada.
Já para investidores mais conservadores, as ações da BRADESCO e RENTACAR apresentam uma combinação equilibrada entre risco e retorno.

Essas conclusões podem orientar a tomada de decisão de investimento, considerando o perfil de risco e os objetivos financeiros de cada investidor.

# Metadados da Análise Financeira

- **Analista de Dados:** Claudio G Vargas
- **Fonte de Dados:** [Yahoo Finance](https://finance.yahoo.com/)
- **Curso de Aquisição de Conhecimento:** IA Expert Academy - Plataforma de Cursos sobre Ciência de Dados e IA
- **Inteligência Artificial Utilizada:** ChatGPT by OpenAI

## Ferramentas Utilizadas

- **Python:** Linguagem de programação utilizada para realizar a análise.
- **Bibliotecas Python:**
  - **Pandas:** Utilizada para manipulação e análise de dados.
  - **Markdown:** Utilizada para formatar e criar a apresentação em formato markdown.
  - **Matplotlib e Seaborn:** Opcionalmente, poderiam ser utilizadas para visualização de dados, caso necessário, mas não foram aplicadas neste caso específico.

***
## Alocação e Otimização de Portfólios


In [12]:
dataset.columns

Index(['Date', 'PETROBRAS', 'VALE', 'BRADESCO', 'RENTACAR', 'BRFOOD', 'BOVA'], dtype='object')

A função `alocacao_ativos` tem o objetivo de simular a alocação de ativos financeiros em um portfólio, distribuindo um valor total de investimento entre diferentes ações com base em pesos definidos aleatoriamente ou fornecidos pelo usuário. Abaixo está uma explicação detalhada do funcionamento da função:

1. **Argumentos da Função**:
   - `dataset`: O DataFrame contendo os dados das ações.
   - `dinheiro_total`: O valor total a ser investido.
   - `seed=0`: Um parâmetro opcional que define a semente para a geração de números aleatórios. Se não for especificado, a semente padrão é 0.
   - `melhores_pesos=[]`: Uma lista opcional de pesos pré-definidos para as ações. Se essa lista estiver vazia, os pesos serão gerados aleatoriamente.

2. **Preparação do Dataset**:
   - Uma cópia do DataFrame `dataset` é criada para evitar alterações no conjunto de dados original.
   - Se a semente (`seed`) for diferente de zero, ela é definida para garantir a reprodutibilidade dos resultados.
   - Se a lista de melhores pesos (`melhores_pesos`) for fornecida, esses pesos são utilizados. Caso contrário, pesos aleatórios são gerados e normalizados para garantir que sua soma seja igual a 1.
   - Os preços das ações no DataFrame são normalizados para começar com o valor 1. Isso permite calcular o retorno em termos percentuais.

3. **Cálculo dos Investimentos Diários**:
   - Para cada ação no DataFrame, os preços são multiplicados pelos pesos correspondentes e pelo valor total investido. Isso calcula o lucro ou prejuízo diário para cada ação, levando em consideração a alocação de pesos.
   
4. **Cálculo do Total Investido**:
   - Uma nova coluna chamada "Total Investido" é adicionada ao DataFrame, que representa o valor total investido em todas as ações em cada dia.

5. **Cálculo da Taxa de Retorno**:
   - Para cada dia no DataFrame, a taxa de retorno é calculada usando a fórmula da taxa de retorno simples. Essa taxa é adicionada como uma nova coluna chamada "Taxa Retorno", representando o retorno percentual em relação ao dia anterior.

6. **Criação do DataFrame de Ações e Pesos**:
   - Um DataFrame é criado para armazenar as ações e seus pesos correspondentes. Isso é útil para análises adicionais e visualizações.

7. **Retorno dos Resultados**:
   - A função retorna o DataFrame atualizado, as datas, o DataFrame de ações com pesos definidos e o valor total investido no último dia.

Essa função é útil para simular a alocação de ativos financeiros em um portfólio, permitindo análises de retorno e risco com base em diferentes estratégias de investimento.

In [15]:
def alocacao_ativos(dataset, dinheiro_total, seed=0, melhores_pesos=[]):
    dataset = dataset.copy()
    if seed != 0:
        np.random.seed(seed)
    if len(melhores_pesos) > 0:
        pesos = melhores_pesos
    else:
        # cria os pesos para cada ação aleatóriamente
        pesos = np.random.random(len(dataset.columns) -1)
        # efetua a normalização dos pesos
        pesos = pesos / pesos.sum()
        #print(f'Percentual dos pesos: {pesos} Total dos pesos: {pesos.sum()}')
    
    # Normalizando os ativos para que todos tenham o valor 1.000 no início
    colunas = dataset.columns[1:]
    for i in colunas:
        dataset[i] = dataset[i] / dataset[i][0]
    
    """
        Através do enumerate vamos percorrer o nome das ações(acao) 
        juntamente com uma variável que irá controlar os indices(i)
        e começando pela: 'colunas = dataset.columns[1:]' até o final.
        Com isso iremos multiplicar o valor de cada ação por dia 
        pelos pesos e pelo valor total aplicado para achar os lucros
        ou prejuísos diários. 
    """
    for i, acao in enumerate(colunas):
        # efetua a multiplicação diária 
        dataset[acao] = dataset[acao] * pesos[i] * dinheiro_total
    
    # cria uma nova coluna no DataFrame e calcula o valor 'total investido'
    # seleciona apenas colunas numéricas
    coluna_numerica = dataset.select_dtypes(include='number')
    dataset['Total Investido'] = coluna_numerica.sum(axis=1)

    # criação do DataFrame somente com as datas
    datas = dataset['Date']  

    # apaga a coluna 'Date' do DataFrame
    dataset.drop(labels=['Date'], axis=1, inplace=True)

    # cria a coluna 'Taxa Retorno' no DataFrame
    dataset['Taxa Retorno'] = 0.0
    
    # efetua o calculo utilizando a taxa de retorno simples adiciona (-1) após a divisão e resultado em percentuais %
    for i in range(1, len(dataset)):
        dataset.loc[i, 'Taxa Retorno'] = ((dataset['Total Investido'][i] / dataset['Total Investido'][i - 1]) -1) * 100

    # cria uma coluna com as ações e outra com os pesos correspondentes
    acoes_pesos = pd.DataFrame(data = {'Ações': colunas, 'Pesos': pesos * 100})

    # Retorna o dataset, as datas, os pesos das ações definidos aleatóriamente, e o ultimo valor da coluna: 'Total Investido'
    return dataset, datas, acoes_pesos, dataset.loc[len(dataset) - 1]['Total Investido'] 

In [16]:
# Efetuando a chamada da função adicionando os argumentos nos parâmetros
dataset, datas, acoes_pesos, soma_valor = alocacao_ativos(pd.read_csv('acoes_novas.csv'), 5000, 10)

In [17]:
dataset

PETROBRAS        VALE     BRADESCO     RENTACAR      BRFOOD  \
0     1330.859777   35.806036  1093.315674  1292.008683  860.138954   
1     1217.111183   35.267597  1094.909516  1238.297720  843.815940   
2     1177.299039   36.680993  1130.928278  1263.313862  851.424122   
3     1232.751581   38.027086  1175.872185  1276.557598  889.465033   
4     1312.375597   38.430914  1181.928445  1254.852537  873.557006   
...           ...         ...          ...          ...         ...   
1421  2826.655152  105.213878  1495.925181  7856.725360  241.110023   
1422  2654.610369  101.394342  1411.250138  7610.936902  235.161816   
1423  2742.765733  104.355747  1397.137657  7612.153607  234.055166   
1424  2693.000621  101.882302  1354.800213  7383.399490  232.118543   
1425  2793.952465  106.576798  1401.169684  7759.383807  236.268465   

            BOVA  Total Investido  Taxa Retorno  
0     387.870876      5000.000000      0.000000  
1     380.156147      4809.558103     -3.808838  
2     382.290029      4841.936322      0.673206  
3     395.175284      5007.848767      3.426572  
4     398.129839      5059.274337      1.026899  
...          ...              ...           ...  
1421  785.262100     13310.891694      0.673691  
1422  751.448534     12764.802101     -4.102577  
1423  761.625471     12852.093382      0.683844  
1424  744.062131     12509.263300     -2.667504  
1425  757.193591     13054.544810      4.359022  

[1426 rows x 8 columns]

In [18]:
datas

0       2015-01-02
1       2015-01-05
2       2015-01-06
3       2015-01-07
4       2015-01-08
           ...    
1421    2020-10-27
1422    2020-10-28
1423    2020-10-29
1424    2020-10-30
1425    2020-11-03
Name: Date, Length: 1426, dtype: object

In [19]:
acoes_pesos

Ações      Pesos
0  PETROBRAS  26.617196
1       VALE   0.716121
2   BRADESCO  21.866313
3   RENTACAR  25.840174
4     BRFOOD  17.202779
5       BOVA   7.757418

>Obs.: valor referente aos ganhos ao final do período

In [20]:
soma_valor

13054.544810388024

## Explorando a Alocação Aleatória de Ativos e os Ganhos no Período

Neste exemplo, além de analisarmos a alocação aleatória de ativos em um portfólio de investimentos, também vamos verificar os ganhos gerados durante o período de tempo considerado.

### Percentual de Pesos

Os pesos representam a proporção de cada ativo no portfólio. Foram gerados aleatoriamente e normalizados para que a soma total seja igual a 100%. Isso significa que cada peso indica a proporção do dinheiro total investido em cada ativo.

## Desempenho do Portfólio ao Longo do Tempo

O DataFrame exibe o desempenho do portfólio ao longo do tempo. Cada linha representa um dia de negociação, começando em 2015-01-02 até 2020-11-03. Para cada dia, são mostrados os valores investidos em cada ativo, bem como o total investido.

## Ganhos no Período

Além dos investimentos diários, agora incluímos os ganhos acumulados durante o período. Isso nos permite ver claramente se houve lucro ou prejuízo ao longo do tempo.

## Interpretação dos Resultados

- Data: Data da observação.
- PETROBRAS, VALE, BRADESCO, RENTACAR, BRFOOD, BOVA: Valor investido em cada ativo naquela data.
- Total Investido: Soma total do valor investido em todos os ativos naquela data.
- Taxa de Retorno: Porcentagem de retorno do investimento em relação ao dia anterior.
- Ganhos Acumulados: Lucro acumulado ao longo do período considerado.

Ao analisar os ganhos acumulados, podemos verificar que, mesmo com a alocação aleatória de ativos, houve períodos de lucro no portfólio. Isso demonstra que, mesmo com uma abordagem inicialmente aleatória, investir em ações pode ser lucrativo ao longo do tempo.

Investir no mercado de ações pode parecer intimidante, mas ao entender como analisar o desempenho do seu portfólio e identificar oportunidades de lucro, você pode tomar decisões financeiras mais informadas e potencialmente lucrativas.

Portanto, convido você a considerar a análise de dados como esta como parte de sua jornada de investimento. Ao entender os padrões e tendências do mercado, você estará mais preparado para tomar decisões que impulsionem seu sucesso financeiro a longo prazo.

Visualização dos dados:

In [24]:
figura = px.line(x= datas, y= dataset['Total Investido'], title= 'Retorno diário do portfólio')
figura.show()

In [25]:
figura = px.line(title= 'Evolução do Patrimônio')
# Acesso as colunas do DataFrame e apagando as colunas: 'Total Investido' e 'Taxa Retorno'
for i in dataset.drop(columns= ['Total Investido', 'Taxa Retorno']).columns:
    figura.add_scatter( x= datas, y= dataset[i], name= i)

figura.show()

In [26]:
figura = px.line(x=datas, y=dataset['Total Investido'], title='Evolução do patrimônio', labels={'x':'Data', 'y':'Valor'})
figura.update_layout(
xaxis_title='Datas',
yaxis_title='Valores',
xaxis_title_font_color='red',  # Cor do título do eixo x
yaxis_title_font_color='blue'   # Cor do título do eixo y
)

figura.show()